lib/recursive-methods.rb

```diff
@@ -1,49 +1,177 @@
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+
+  reversed_s = reverse(s[1..-1])
+  reversed_s += s[0]
+  return reversed_s
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Reverse in place
+
+# Write a method `reverse_inplace` that accepts a string as a parameter and then reverses the string in place using a recursive algorithm.
+# - e.g. reverse("hello world") will convert the input string to "dlrow olleh"
+
+# From example in class
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return reverse_help(s, 0, s.length - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def reverse_help(s, first = 0, last = s.length - 1)
+  if first < last
+    s[first], s[last] = s[last], s[first]
+    return reverse_help(s, first + 1, last - 1)
+  end
+
+  return s
+end
+
+# Bunny ears
+
+# Write a method `bunny` that accepts an integer parameter n. N represents a number of bunnies and each bunny has two big floppy ears. We want to compute the total number of ears across all the bunnies recursively (without loops or multiplication).
+# - e.g. bunny(0) = 0
+# - e.g. bunny(1) = 2
+# - e.g. bunny(10) = 20
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n == 0
+  return 2 + bunny(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Nested
+
+# Write a method `nested` that accepts a string of only parenthesis and then returns if those parenthesis are properly nested. You may
+# assume that no non-parenthesis characters will be passed to this method.
+# - e.g. nested("((()))") = true
+# - e.g. nested("())") = false
+# **Challenge**: Try doing this without creating new strings in the process of solving the problem.
+
+# Time complexity: O(n)
+# Space complexity: o(n^2)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length.odd?
+    return false
+  elsif s == ""
+    return true
+  end
+
+  return nested_help(s, 0, s.length - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+def nested_help(s, first, last)
+  if first >= last
+    return true
+  elsif s[first] == "(" && s[last] == ")"
+    return nested_help(s, first + 1, last - 1)
+  else
+    return false
+  end
+
+  return s
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Search
+
+# Write a method `search` that accepts an unsorted array of integers and an integer value to find and then returns true if the value is found in the unsorted array and returns false otherwise. Make the algorithm recursive.
+# - e.g. search([34, 45, 21, 4, 67], 4) should return true
+# - e.g. search([24, 51, 11], 4) should return false
+# - e.g. search([], 4) should return false
+
+# Time complexity: 0(n)
+# Space complexity: O(n^2)
+def search(array, value, index = 0)
```

Comment:
You're not creating a new array with each recursive call so the space complexity is O(n)